code/modules/implant_hacking/_hacking.dm

```diff
@@ -0,0 +1,167 @@
+		cleanup()
+		return
+
+	if(!game_list.len)
+		for(var/i in 1 to diffrences)
+			var/datum/hacking_minigame/game = new/datum/hacking_minigame()
+			game.generate()
+			game_list += game
+
+	ui_interact(user)
+
+/obj/item/cyberlink_connector/proc/cleanup()
+	current_user = null
+	cybernetic = null
+	QDEL_LIST(game_list)
+	processing = FALSE
+
+/obj/item/cyberlink_connector/proc/hack_success(success as num)
+	for(var/info in cybernetic.encode_info)
+		if(cybernetic.encode_info[info] == NO_PROTOCOL)
+			continue
+		//Not a += because we want to avoid having duplicate entries in either encode_info
+		cybernetic.encode_info[info] |= parent_cyberlink.encode_info[info]
+	current_user.mind.adjust_experience(/datum/skill/implant_hacking,success * 25)
+	to_chat(current_user,"<span class='notice'> Cyberlink beeps: HACKING [uppertext(cybernetic.name)] SUCCESS. COMPATIBILITY ACHIEVED.</span>")
+	cleanup()
+
+
+/obj/item/cyberlink_connector/proc/hack_failure(failed as num)
+	var/chance = rand(0,40*failed)
+	switch(chance)
+		if(0 to 25)
+			to_chat(current_user,"<span class='warning'> Cyberlink beeps: HACKING [uppertext(cybernetic.name)] MINOR FAILURE. COMPATIBILITY NOT ACHIEVED. NO DAMAGE DETECTED.</span>")
+		if(26 to 40)
+			to_chat(current_user,"<span class='warning'> Cyberlink beeps: HACKING [uppertext(cybernetic.name)] MEDIUM FAILURE. COMPATIBILITY NOT ACHIEVED. SMALL AMOUNT OF DAMAGE DETECTED.</span>")
+			current_user.adjustFireLoss(10)
+			current_user.emote("scream")
+		if(41 to 50)
+			to_chat(current_user,"<span class='warning'> Cyberlink beeps: HACKING [uppertext(cybernetic.name)] MEDIUM FAILURE. COMPATIBILITY NOT ACHIEVED. PROTOCOL SCRAMBILING DETECTED.</span>")
+			cybernetic.random_encode()
+		if(51 to 75)
+			to_chat(current_user,"<span class='danger'> Cyberlink beeps: HACKING [uppertext(cybernetic.name)] MAJOR FAILURE. COMPATIBILITY NOT ACHIEVED. MINOR ELECTROMAGNETIC PULSE DETECTED.</span>")
+			empulse(current_user, 0, 1)
+		if(76 to 99)
+			to_chat(current_user,"<span class='danger'> Cyberlink beeps: HACKING [uppertext(cybernetic.name)] MAJOR FAILURE. COMPATIBILITY NOT ACHIEVED. MAJOR ELECTROMAGNETIC PULSE DETECTED.</span>")
+			empulse(current_user, 1, 2)
+		if(100 to INFINITY)
+			to_chat(current_user,"<span class='danger'> Cyberlink beeps: HACKING [uppertext(cybernetic.name)] CRITICAL FAILURE. COMPATIBILITY NOT ACHIEVED. IMPLANT OVERHEATING IN 5 SECONDS.</span>")
+			cybernetic.visible_message("<span class='danger'>[cybernetic.name] begins to flare and twitch as the electronics fry and sizzle!</span>")
+			addtimer(CALLBACK(src, .proc/explode), 5 SECONDS)
+	current_user.mind.adjust_experience(/datum/skill/implant_hacking,(4 - failed)*2)
+	cleanup()
+
+/obj/item/cyberlink_connector/proc/explode()
+	SIGNAL_HANDLER
+
+	dyn_explosion(get_turf(cybernetic),2,1)
+	qdel(src)
+
+/obj/item/cyberlink_connector/process(delta_time)
```

Comment:
This should really just be event based, is there any need for process() here that I'm missing?

Reply:
for the counter time down, i made it so it doesn't use a timer but a variable ticking down, either way it would perform exactly the same. i dont see the reason for changing it.

Reply:
scratch that, changed my mind gonna work on it

Reply:
made it event based